src/back/glsl_common.rs

```diff
@@ -0,0 +1,704 @@
+                    ScalarKind::Float => match width {
+                        4 => "",
+                        8 => "d",
+                        _ =>
+                            return Err(Error::Custom(format!(
+                                "Cannot build float of width {}",
+                                width
+                            ))),
+                    },
+                    ScalarKind::Bool => "b",
+                },
+                *size as u8
+            ),
+            TypeInner::Matrix {
+                columns,
+                rows,
+                kind,
+                width,
+            } => format!(
+                "{}mat{}x{}",
+                match kind {
+                    ScalarKind::Sint => "i",
+                    ScalarKind::Uint => "u",
+                    ScalarKind::Float => match width {
+                        4 => "",
+                        8 => "d",
+                        _ =>
+                            return Err(Error::Custom(format!(
+                                "Cannot build float of width {}",
+                                width
+                            ))),
+                    },
+                    ScalarKind::Bool => "b",
+                },
+                *columns as u8,
+                *rows as u8
+            ),
+            TypeInner::Pointer { base, class } => todo!(),
+            TypeInner::Array { base, size, stride } => format!(
+                "{}[{}]",
+                base.write_glsl(types, structs)?,
+                size.write_glsl()
+            ),
+            TypeInner::Struct { .. } => structs.get(self).unwrap().0.clone(),
+            TypeInner::Image { base, dim, flags } => format!(
+                "{}texture{}{}",
+                match types[*base].inner {
+                    TypeInner::Scalar { kind, .. } => match kind {
+                        ScalarKind::Sint => "i",
+                        ScalarKind::Uint => "u",
+                        ScalarKind::Float => "",
+                        _ => return Err(Error::Custom(format!("Cannot build image of booleans",))),
+                    },
+                    _ =>
+                        return Err(Error::Custom(format!(
+                            "Cannot build image of type {}",
+                            base.write_glsl(types, structs)?
+                        ))),
+                },
+                dim.write_glsl(),
+                flags.write_glsl()
+            ),
+            TypeInner::DepthImage { dim, arrayed } => format!(
+                "texture{}{}",
+                dim.write_glsl(),
+                if *arrayed { "Array" } else { "" }
+            ),
+            TypeInner::Sampler { comparison } => String::from(if *comparison {
+                "sampler"
+            } else {
+                "samplerShadow"
+            }),
+        })
+    }
+}
+
+impl StorageClass {
+    pub(crate) fn write_glsl<'a>(&'a self) -> impl fmt::Display + 'a {
```

Comment:
same here. I understand this is a matter of preference. My confusion comes from the fact that `self` is just a piece of data here. It feels like it's abusing the "method" semantic on types, which makes more sense for types that have actual contents in them, and meant to have associated logic.